[stdlib] Fix: Switch `TestReport.name` from `StaticString` to `String`

josiahls opened the pull request:
This is to support runtime test discovery.

Example runtime test discovery:
```mojo
from subprocess import run
from testing.suite import TestSuiteReport, TestReport, TestResult
from builtin._location import __call_location
from pathlib import Path
from time import perf_counter_ns

comptime TEST_DIR = Path("tests")

def test_file(file: Path) -> TestReport:
    var result = run("pixi run test " + String(file))
    if "Unhandled exception caught during execution" in result:
        return TestReport.failed(
            name=file.name(), duration_ns=0, error=Error(result)
        )
    return TestReport.passed(name=file.name(), duration_ns=0)

def walk_tests(path: Path, mut test_results: List[TestReport]):
    for f in path.listdir():
        file = path / f
        if file.is_file() and file.suffix() == ".mojo":
            var result = test_file(file)
            test_results.append(result^)
        elif file.is_dir():
            walk_tests(file, test_results)


def main():
    var test_results = List[TestReport]()
    walk_tests(TEST_DIR, test_results)
    var report = TestSuiteReport(
        reports=test_results^, location=__call_location[inline_count=0]()
    )
    print(report)
```
Compile time inter-file test discover also still works:
```mojo
from testing import assert_equal, TestSuite

def test_lightbug_http_backend():
    assert_equal(2, 1)

def main():
    TestSuite.discover_tests[__functions_in_module()]().run()
```
Example output:
<img width="1004" height="334" alt="Screenshot 2025-11-26 at 3 43 29 PM" src="https://github.com/user-attachments/assets/c6bf47e6-fc9e-4a93-ade9-178543a16b01" />

josiahls committed:
[stdlib] Fix: Switch `TestReport.name` from StaticString to String

This is to support runtime test discovery.

NathanSWard commented on mojo/stdlib/stdlib/testing/suite.mojo:
```diff
@@ -151,9 +151,7 @@ struct TestReport(Copyable, Movable, Writable):
         }
```
This function (also `failed` and `skipped`) should take a `var String` and then move the string to avoid unnecessary copies.

```mojo
return {
    name = name^,
    duration_ns = duration_ns,
    result = TestResult.PASS,
}
```

josiahls replied:
done

NathanSWard commented on mojo/stdlib/stdlib/testing/suite.mojo:
```diff
@@ -187,7 +185,7 @@ struct TestReport(Copyable, Movable, Writable):
         out self,
         *,
-        name: StaticString,
+        name: String,
```
Ditto `var String` and then later moving the string 
```mojo
self.name = name^
```

josiahls replied:
done

soraros commented on mojo/stdlib/stdlib/testing/suite.mojo:
```diff
@@ -397,7 +395,7 @@ struct TestSuite(Movable):
         self.tests = List[_Test]()
         self.location = location.or_else(__call_location())
-        self.skip_list = Set[StaticString]()
+        self.skip_list = Set[String]()
```
```suggestion
        self.skip_list = {}
```

josiahls replied:
done

josiahls committed:
docs: Remove TODO in suite.mojo regarding preferring StaticString over String

`TestSuite` no longer restricts itself to just
compile time test orchestration, but instead supports
comptime and runtime test orchestration.
